File: java/TuduManager/web.api.spring.boot/src/test/java/net/seymourpoler/tudumanager/web/api/spring/boot/todo/update/UpdateTodoControllerShould.java
package net.seymourpoler.tudumanager.web.api.spring.boot.todo.update;

import net.seymourpoler.tudumanager.domain.ErrorCodes;
import net.seymourpoler.tudumanager.domain.ServiceExecutionResult;
import net.seymourpoler.tudumanager.domain.todo.update.IUpdateTodoService;
import net.seymourpoler.tudumanager.domain.todo.update.TodoUpdatingRequest;
import org.junit.Before;
import org.junit.Test;
import org.springframework.http.HttpStatus;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class UpdateTodoControllerShould {
    IUpdateTodoService updateTodoService;
    UpdateTodoController controller;

    @Before
    public void setUp(){
        updateTodoService = mock(IUpdateTodoService.class);
        controller = new UpdateTodoController(updateTodoService);
    }

    @Test
    public void
    return_not_found_when_is_not_found(){
        var request = new HttpTodoUpdatingRequest();
        var errors = List.of(new net.seymourpoler.tudumanager.domain.Error("email", ErrorCodes.NotFound));
        var errorResult = ServiceExecutionResult.of(errors);
        when(updateTodoService.update(any())).thenReturn(errorResult);

        var response = controller.update(request);

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.NOT_FOUND);
    }

    @Test
    public void
    return_bad_request_when_there_is_an_error(){
        var request = new HttpTodoUpdatingRequest();
        var errors = List.of(new net.seymourpoler.tudumanager.domain.Error("email", ErrorCodes.Required));
        var errorResult = ServiceExecutionResult.of(errors);
        when(updateTodoService.update(any())).thenReturn(errorResult);

        var response = controller.update(request);

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.BAD_REQUEST);
        assertThat(response.getBody()).isEqualTo(errors);
    }
}

File: dotnet/react.mvc-core.postgres.dapper/TuduManager/test/TuduManayer.web.api.Test/Todo/Update/UpdateTodoControllerShould.cs
using System.Collections.Generic;
using System.Net;
using Microsoft.AspNetCore.Mvc;
using Moq;
using Shouldly;
using TuduManayer.Domain;
using TuduManayer.Domain.Todo;
using TuduManayer.web.api.Todo.Update;
using Xunit;

namespace TuduManayer.web.api.Test.Todo.Update
{
    public class UpdateTodoControllerShould
    {
        [Fact]
        public void return_error_when_there_are_errors()
        {
            const string title = "title";
            var service = new Mock<IUpdateTodoService>();
            service
                .Setup(x => x.Update(It.Is<TodoUpdatingArgs>(y => y.Title == title)))
                .Returns(ServiceExecutionResult.WithErrors(new List<Error>{Error.With(nameof(TodoUpdatingArgs.Title), ErrorCodes.Required)}));
            var controller = new UpdateTodoController(service.Object);
            var request = new TodoUpdatingRequest{Id = 1, Title = title};
            
            var response = controller.Update(request) as BadRequestObjectResult;
            
            response.StatusCode.ShouldBe((int)HttpStatusCode.BadRequest);
        }

        [Fact]
        public void return_ok()
        {
            const int someId = 1;
            var service = new Mock<IUpdateTodoService>();
            service
                .Setup(x => x.Update(It.Is<TodoUpdatingArgs>(y => y.Id == someId)))
                .Returns(ServiceExecutionResult.WithSucess());
            var controller = new UpdateTodoController(service.Object);
            var request = new TodoUpdatingRequest{Id = someId, Title = "some title", Description = "some description"};
            
            var response = controller.Update(request) as OkResult;
            
            response.StatusCode.ShouldBe((int)HttpStatusCode.OK);
        }
    }
}